### DISCUSSION TITLE
Composite id with UnsavedValue="any" disallows to delete entity

### DISCUSSION THREAD

**Discussion by user2291296** [nhibernate/nhibernate-core#2725] [Q&A]
We use a table with a composite primary key for many-to-many relationship. The mapping is shown below:

`mapping.CompositeId().KeyReference(e => e.POINT).KeyReference(e => e.NODE).UnsavedValue("any");`

To prevent unnecessary SELECT sql-commands in SaveOrUpdate we added UnsavedValue("any") option in mapping.
But after that we noticed that entity deletion command does not delete entity any more:

`session.Delete(entity);`

There is neither exception nor warning. The command is just ignored.
If we remove `UnsavedValue("any")` from mapping then `session.Delete(entity)` works perfectly, but we get big number of excessive selects during saving entities.

Is there any way to delete entity with composite id and UnsavedValue=any?

Version of NHibernate is 5.3.8.
Database is Oracle.

**Comment by fredericDelaporte** [comment]
This is by design.

When using `Delete(entity)` with an object not attached into the session, the session checks if it is considered transient or not.
If you tell NHibernate that any identifier value means the entity is unsaved (so transient), then the session considers it transient, so nothing needs to be actually deleted in database.

> To prevent unnecessary SELECT sql-commands in SaveOrUpdate we added UnsavedValue("any") option in mapping.

This should also have the "side effect" of causing NHibernate to always insert the entities even when they already exist, when doing a `SaveOrUpdate`.  With assigned identifier, if your code does not know whether the entity already exist or not, you cannot avoid the select. If your code knows, make it use the adequate method: `Update` if the provided detached entity is supposed to exist, `Save` if it is transient. (If it is already attached to the session, just `Flush`/`Commit` the session of course, no need to call neither `Update` nor `Save` in such case.)

Or you should cease working directly on these entities but instead build [DML statements](https://nhibernate.info/doc/nhibernate-reference/batch.html#batch-direct).

**Reply by bahusoid** [reply]
> When using Delete(entity) with an object not attached into the session, the session checks if it is considered transient or not.

So if you need to delete object by id use `session.Load`:
```C#
session.Delete(session.Load<Entity>(id));
```

**Reply by user2291296** [reply]
Suggested statement (see below) adds unnecessary database communication.
> session.Delete(session.Load<Entity>(id));

If many entities must be deleted then it may lead to bad performance.

**Reply by user2291296** [reply]
> This is by design.

OK. So `UnsavedValue("any")` forces NHibernate always to execute INSERT sql-statement on `SaveOrUpdate` regardless of cache.

@fredericDelaporte , is there any mapping option which forces NHibernate to execute DELETE sql-statement on `Delete(entity)` regardless of cache?

**Reply by bahusoid** [reply]
> Suggested statement (see below) adds unnecessary database communication.

Why? Is proxy disabled? `session.Load` doesn't trigger entity loading. It returns proxy.

**Reply by fredericDelaporte** [reply]
> Suggested statement (see below) adds unnecessary database communication.
>
>>   session.Delete(session.Load(id));
>
>If many entities must be deleted then it may lead to bad performance.

`Load` does not trigger a select in database, unless you have disabled proxy. Load just yield back a proxy, without checking if the entity exists or not. Its semantic is to assume the asked entity exist.
But with an entity having a composite id, it can only work if you have [mapped the id as a component](https://nhibernate.info/doc/nhibernate-reference/components.html#components-compositeid), which is recommended.

By the way, you should read the documentation about [assigned identifiers](https://nhibernate.info/doc/nhibernate-reference/mapping.html#mapping-declaration-id-assigned). That does apply to composite ids too, which are a special kind of assigned identifiers. And it tells you all about the behavior you are currently asking about:
> Due to its inherent nature, entities that use this generator may, when saved via the ISession's SaveOrUpdate() method, cause NHibernate to query the database for checking if the entity exist or not. 
>
> To avoid this, either:

Read more by following above links.

**Reply by bahusoid** [reply]
> But with an entity having a composite id, it can only work if you have mapped the id as a component,

Ah.. Indeed. In this case it seems associating entity with session before deletion via `session.Lock` should work:
```C#
session.Lock(entity, LockMode.None);
session.Delete(entity);
```
